dtbase/services/base.py

```diff
@@ -0,0 +1,218 @@
+        Method for getting data from a source and returning Backend API Endpoints names
+        and payload pairs. This method should be implemented when inheriting from this
+        class. The implementation in BaseService only raises a NotImplementedError.
+
+        The method should return a list of tuples. A tuple should
+        be in the format [(<endpoint_name>, <payload>)]. It must be a list even if
+        its a single tuple.
+
+        There is more information in the docstrings of the BaseIngress and BaseModel
+        get_service_data methods about the expected format of the tuples.
+        """
+        raise NotImplementedError(
+            "The user should implement this method"
+            " by inheriting from the BaseService class"
+            " and overwriting the get_service_data method."
+        )
+
+    def post_service_data(
+        self,
+        data_pairs: List[tuple],
+        dt_user_email: Optional[str] = None,
+        dt_user_password: Optional[str] = None,
+    ) -> List[Response]:
+        """
+        Upload data to the database via the backend via a post request.
+        Handles unpacking of data pairs and returns responses from the backend.
+
+        Args:
+            data_pairs: list of tuples of the form [(<endpoint_name>, <payload>)].
+            It must be a list even if its a single tuple. Typically, this is the output
+            from self.run().
+            dt_user_email: email of the backend user to login with. By default read from
+                the environment variable DT_DEFAULT_USER_EMAIL.
+            dt_user_password: password of the backend user to login with. By default
+                read from the environment variable DT_DEFAULT_USER_PASS.
+
+        Returns:
+            List of responses from the backend API calls.
+        """
+        if dt_user_email is None:
+            dt_user_email = DEFAULT_USER_EMAIL
+        if dt_user_password is None:
+            dt_user_password = DEFAULT_USER_PASS
+
+        self._backend_login(dt_user_email, dt_user_email)
+
+        responses = []
+        for data_pair in data_pairs:
+            endpoint, payload = data_pair
+            response = auth_backend_call(
+                "post", endpoint, payload=payload, token=self.access_token
+            )
+            log_rest_response(response)
+            responses.append(response)
+        return responses
+
+    def __call__(
+        self,
+        dt_user_email: Optional[str] = None,
+        dt_user_password: Optional[str] = None,
+        **kwargs: Any,
+    ) -> List[Response]:
+        """
+        This method will 'call' or run the service.
+        It runs the get_service_data and post_service_data methods in sequence.
+
+        Args and Kwargs are passed to the get_service_data method.
+        """
+        data_pairs = self.get_service_data(**kwargs)
+        return self.post_service_data(data_pairs, dt_user_email, dt_user_password)
+
+    def schedule(self) -> None:
```

Comment:
I think this logic may eventually end up living somewhere else. I think `BaseService` is essentially what someone implementing a service will subclass, to interface with DTBase. I think the scheduling logic is likely to instead live in the backend. Or to put this another way `BaseService` would deal with "what happens when someone asks this service to run" and the backend would deal with "when should each service be run".

Reply:
Agreed, I'll remove it for now as it isn't implemented